Gorry's ECN rewrite

martinthomson opened the pull request:
This contains a lot of changes, but these seem to all be editorial.

@gorryfair is unable to engage here for a little while, but we should take that opportunity to provide feedback.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
 increases to ECT(1) and ECN-CE counts is less than the number of newly
 acknowledged packets sent with an ECT(1) marking.  These checks can detect
-removal of ECN markings in the network.
+remarking of ECN CE-markings by the network.
```
I know that "remarking" is likely a term of art, but it is also a word with a completely different meaning.  I think we should avoid it.

janaiyengar replied:
"These checks can detect network elements that overwrite ECN-CE markings with an ECT(0) or ECT(1) codepoint."

gorryfair replied:
I'm OK with a change of word, but the checks only detect some changes to the ECN-Field. They can't detect rewriting of a CE-Marking back to ECT(0) (for instance when ECT(0) was sent), so we should be careful to avoid giving that impression..

martinthomson reviewed (commented):
This is a pretty disruptive and difficult to review.  I will need to take another pass at this.

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
 
-### ECN Validation {#ecn-validation}
+Each packet number space maintains separate acknowledgement state and separate
+ECN counts.  Coalesced packets (see {{packet-coalesce}}) carry several QUIC
+packets that share the same IP header.  The ECN count for the ECN codepoint
```
Perhaps "coalesced QUIC packets". Also "ECN counts" (plural, since different PN spaces).

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
-successfully processed ACK frame. Any increase in ECN counts is validated
-based on the markings that were applied to packets that are newly
-acknowledged in the ACK frame.
+Erroneous application of CE-marks by the network can result in degraded
```
```suggestion
Erroneous application of ECN-CE markings by the network can result in degraded
```

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
 increases to ECT(1) and ECN-CE counts is less than the number of newly
 acknowledged packets sent with an ECT(1) marking.  These checks can detect
-removal of ECN markings in the network.
+remarking of ECN CE-markings by the network.
```
```suggestion
remarking of ECN-CE markings by the network.
```

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
-present in the ACK frame.  This check detects a network element that zeroes out
-ECN bits or a peer that is unable to access ECN markings.
+If an ACK frame newly acknowledges a packet that the endpoint sent setting
+either the ECT(0) or ECT(1) codepoint, ECN validation fails if the ECN counts
```
```suggestion
either the ECT(0) or ECT(1) codepoint, ECN validation fails if the corresponding ECN counts
```

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
-newly acknowledged packets.
+It is therefore possible for the total increase in the ECT(0), ECT(1), and
+ECN-CE counts to be greater than the number of packets acknowledged in an ACK
+frame.  This is why ECN counts are permitted to be larger than the value
+corresponding to the largest acknowledged packet number.
```
The old text makes sense to me, but I don't understand this proposed text. What value corresponding to the largest acknowledged PN is this text talking about?

gorryfair replied:
(I couldn't clearly understand the old text in the old context... but expect the intention was fine.) Could someone propose a new sentence following the structure of the old one?

gorryfair replied:
I think the original text was this - I wouldn't object to using these words, but I think they could be clearer: "This is why counts are permitted to be larger than might be accounted for by newly acknowledged packets."

mirjak replied:
Didn't we also discuss a check were all packets get marked as CE (including non-ECT packets)? That's actually the only check that is basically implemented today in TCP (checking if the SYN is CE which should be non-ECT).

gorryfair replied:
@mirjak - Can you suggest what text you think is missing, so we can complete the edit?

mirjak replied:
This might be more than an edit but I guess there could be another case, that checks e.g. after one RTT if all packets got CE marked and then fails validation.

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
+sending rate in response, as described in {{QUIC-RECOVERY}}.
+
+To enable ECN, a sending QUIC endpoint first determines whether a path supports
+ECN marking and whether reports the ECN value in received IP headers; see
```
```suggestion
ECN marking and whether the peer reports the ECN values in received IP headers; see
```

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
-errors are detected.
+For example, if one each of an Initial, 0-RTT, Handshake, and 1-RTT QUIC packet
+are coalesced, the corresponding counts for the Initial and Handshake packet
+number space will be incremented by one and the counts for the application data
```
```suggestion
number spaces will be incremented by one each and the counts for the application data
```

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
 
-#### Sending ECN Markings
+It is possible for faulty network devices to corrupt or erroneously drop packets
+that set an ECN codepoint.  To provide robust connectivity in the presence of
```
```suggestion
that carry an ECN codepoint.  To provide robust connectivity in the presence of
```

gorryfair replied:
Doesn't every IP packet carry an ECN code point. Maybe we need to say ECT(0) or ECT(1) code point then?

larseggert replied:
Or a non-zero ECN codepoint?

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
 
-* Set the ECT(0) codepoint in the IP header of early outgoing packets sent on a
-  new path to the peer ({{!RFC8311}}).
+* The endpoint SHOULD set the ECT(0) codepoint in the IP header of early
```
```suggestion
* The endpoint MUST set the ECT(0) codepoint in the IP header of early
```
Can't validate if you don't set it.

larseggert replied:
That's fine, too.

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
 
-* If all packets that were sent with the ECT(0) codepoint are eventually deemed
-  lost (Section 6 of {{QUIC-RECOVERY}}), validation is deemed to have failed.
+* The endpoint SHOULD monitor whether all packets sent with an ECT codepoint are
```
```suggestion
* The endpoint MUST monitor whether all packets sent with an ECT codepoint are
```
Can't validate if you don't monitor.

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
-* If all packets that were sent with the ECT(0) codepoint are eventually deemed
-  lost (Section 6 of {{QUIC-RECOVERY}}), validation is deemed to have failed.
+* The endpoint SHOULD monitor whether all packets sent with an ECT codepoint are
+  eventually deemed lost (Section 6 of {{QUIC-RECOVERY}}), deeming this as an
+  indication that ECN validation has failed.
```
```suggestion
  eventually deemed lost (Section 6 of {{QUIC-RECOVERY}}), indicating
  that ECN validation has failed.
```

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
-faulty network element, an endpoint could set the ECT(0) codepoint for only the
-first ten outgoing packets on a path, or for a period of three RTTs, whichever
-occurs first.
+To reduce the chances of misinterpreting loss of packets dropped by a faulty
+network element, an endpoint could set an ECT codepoint for only the first ten
+outgoing packets on a path, or for a period of three RTTs, whichever occurs
+first.
```
More needs to be said here about why this might be beneficial.

gorryfair replied:
This mechanism was there already, so it wasn't my idea. When I edited,  I didn't see where it did hard - nor did I see benefit - so is it needed?

gorryfair replied:
@mirjak - What do you suggest (remove this text)?

mirjak replied:
I argued in another discussion already that we should just recommend to mark all packets as ECT(0) until validation fails as this case where packet get dropped is basically not existence anymore. However, this was not accepted as people felt this approach would not increase implementation complexity. I still think the text should note that this is a really rare case and I had an editorial PR for that that was also not expected, so I would propose anything again. Just wanted to state my opinion...

gorryfair replied:
I think we all need to find text we can agree upon. Here is a proposal:

"If an endpoint has cause to expect that packets setting an ECT codepoint are being dropped by faulty network element, the endpoint could set an ECT codepoint for only the first ten outgoing packets on a path, or for a period of three RTTs. If an endpoint then discovers all non-zero ECN codepoints fail to be delivered, it can assume the path did not support ECN.

Other methods of probing paths for ECN support are possible,..."

mirjak replied:
Looks good to me!

mirjak replied:
I don't think your rewording did change anything from the original text.

mirjak replied:
Sorry, I think your wording is fine. I looked at this text early in the same section which still seems to indicate that dropping is a real problem:

```
It is possible for faulty network devices to corrupt or erroneously drop
 packets that carry a non-zero ECN codepoint. To ensure connectivity in the
 presence of such devices, an endpoint validates the ECN counts for each network
 path and disables use of ECN on that path if errors are detected.
```

gorryfair replied:
I agree,  and suggest this sentence can now be removed, and is now covered by the clarified later sentence: "It is possible for faulty network devices to corrupt or erroneously drop packets that carry a non-zero ECN codepoint."

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
 marking strategies. Implementations MAY use other methods defined in RFCs; see
-{{?RFC8311}}. Implementations that use the ECT(1) codepoint need to perform ECN
-validation using ECT(1) counts.
+{{?RFC8311}}. Implementations that use the ECT(1) codepoint need to
+perform ECN validation using the reported ECT(1) counts.
```
This whole ECN text needs to make up its mind on whether it wants to specify that ECT(0) should be used, or if it wants to allow ECT(0) or ECT(1). At the moment, it's a mix. And if it wants to allow both, some words need to be added whether an implementation should pick one statically or randomly or whatnot.

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
-particular, an endpoint that never applies a particular marking can fail
-validation when a non-zero count for the corresponding marking is received.
-This check can detect when packets are marked ECT(0) or ECT(1) in the network.
+Out of order processing of the ECN counts can result in a validation failure.
```
What does "out of order processing of the ECN counts" mean?

mirjak commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
-#### Sending ECN Markings
+It is possible for faulty network devices to corrupt or erroneously drop packets
+that set an ECN codepoint.  To provide robust connectivity in the presence of
+such devices, an endpoint validates the ECN counts for each network path  and
```
This needs more explanation because you say first something about "drops" and the "validate ECN counts" but to check for drops you need to detect loss of ECT marked packet which is not really discussed here.

martinthomson replied:
This is introduction.  The explanation you seek follows.

mirjak replied:
I just thought the two sentences don't fit together because the first talks about corrupt and drop but the second sentence only talk about validating counts and not about checking on loss. This is an editorial comment.

larseggert commented:
Who holds the pen on this PR?

gorryfair commented:
Martin, (cc Lars)

I think, if I understand, that we now have text that is "mostly" 
correctable from the comments. I supect the proposed changes fix most 
(all?) of the confusion I originally had.  You own the edit token.

On 07/09/2020 12:27, Lars Eggert wrote:
>
> Who holds the pen on this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/quicwg/base-drafts/pull/4059#issuecomment-688263251>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABYLLEXJ7CRAILXKLKVZY4LSES7Q3ANCNFSM4QMZAVOA>.
>
* I don't think I have edit permission to fix the PR text where the PR 
comments propose a change, but you likely can edit!  - or you can make 
this possible for me to do, and I will re-edit these in.

* I think Mirja has a few remaining issues, which I think I'd sympathise 
with, but talking off-list I think the resolution of the comments in the 
PR edit may go a long way to alaso address these changes.

I don't know how best to progress, should we try to land the ECN rewrite 
making all agreed corrections, and then review what is left?

Gorry

mirjak commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,136 @@ The details of loss detection and congestion control are described in
-If validation fails, then the endpoint stops sending ECN markings in subsequent
-IP packets with the expectation that either the network path or the peer does
-not support ECN.
+If validation fails, then the endpoint MUST disable ECN. It stops setting the
```
I don't think this should be a MUST. If normative language is needed I guess it should be a SHOULD with a warning that remarking by the network can conceal congestion. I would however rather simply not use normative language here.

gorryfair replied:
I think it is a MUST, until the next round of validation starts. The endpoint shouldn't continue to set ECT(?) and not be able to respond when CE is marked.

janaiyengar replied:
This is what the original text said, so I think the intent is that of a MUST.

mirjak commented on draft-ietf-quic-transport.md:
```diff
@@ -3838,143 +3838,135 @@ The details of loss detection and congestion control are described in
+For example, if one each of an Initial, 0-RTT, Handshake, and 1-RTT QUIC packet
+are coalesced, the corresponding counts for the Initial and Handshake packet
+number space will be incremented by one and the counts for the application data
+packet number space will be increased by two.
```
That was my question: do we specify correctly that the sender has to correct for double-counting?

martinthomson commented:
Plan here is to have Ian review this and merge it for -30 unless there is something that comes up.

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,136 @@ The details of loss detection and congestion control are described in
-faulty network element, an endpoint could set the ECT(0) codepoint for only the
-first ten outgoing packets on a path, or for a period of three RTTs, whichever
-occurs first.
+An endpoint thus attempts to use ECN and validates this for each new connection,
+when switching to a server's preferred address, and on active connection
+migration to a new path.  {{ecn-alg}} describes one possible algorithm.
```
```suggestion
An endpoint negotiates ECN for each new connection or new path.
{{ecn-alg}} describes one possible algorithm.
```

martinthomson replied:
I don't want to say negotiates.  And I'm not sure that we need the extreme economy you propose otherwise.

gorryfair replied:
I also think the original is much clearer.

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,136 @@ The details of loss detection and congestion control are described in
 
-ECN validation fails if the sum of the increase in ECT(0) and ECN-CE counts is
-less than the number of newly acknowledged packets that were originally sent
+ECN validation also fails if the sum of the increase in ECT(0) and ECN-CE counts
```
undo

```suggestion
ECN validation also if the sum of the increase in ECT(0) and ECN-CE counts
```

janaiyengar replied:
What are you undo'ing?

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,136 @@ The details of loss detection and congestion control are described in
 increases to ECT(1) and ECN-CE counts is less than the number of newly
 acknowledged packets sent with an ECT(1) marking.  These checks can detect
-removal of ECN markings in the network.
+remarking of ECN-CE markings by the network.
```
remarking and removal are completely different and I believe removal is correct.  I'm unsure what's meant by remarking.

```suggestion
removal of ECN-CE markings by the network.
```

janaiyengar replied:
Remarking means that the network marks the bits to something else. The point here is that something in the  network marked the bits to CE, and now another element is remarking it to something else. Perhaps "erasing of ECN-CE markings"?

ianswett replied:
Something to indicate the bits were cleared is important, as the protocol can detect marking of packet that shouldn't be marked in most cases.

ianswett replied:
If I'm reading correctly, the preceding two sentences describe how to detect if ECN bits have been set to 0.  Detecting remarking would indicate to me that you can also detect a bit that was 0 set to 1, but the next paragraph explains why it's ok for the counts to be larger than expected.

That being said, it sounds like people like the word 'remarked' when talking about ECN, so I'm fine to keep it as is.

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,136 @@ The details of loss detection and congestion control are described in
-This is why counts are permitted to be larger than might be accounted for by
-newly acknowledged packets.
+counts to be greater than the number of packets that are newly acknowledged by
+an ACK frame. This is why ECN counts are permitted to be larger than the value
+corresponding to the largest acknowledged packet number.
```
This is unfortunately incorrect.  One could in theory start with a packet number of 1 million and it should not affect correctness.

```suggestion
an ACK frame. This is why ECN counts are permitted to be larger than the total
number of acknowledged packets within a packet number space.
```

martinthomson replied:
How about:

```suggestion
an ACK frame. This is why ECN counts are permitted to be larger than the total
number of packets that are acknowledged.
```
I don't think we need to say "packet number space" again here.

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,136 @@ The details of loss detection and congestion control are described in
-Processing ECN counts out of order can result in validation failure.  An
-endpoint SHOULD skip ECN validation on an ACK frame that does not increase the
-largest acknowledged packet number.
+ECN validation can fail if the received total count for either ECT(0) or ECT(1)
```
Can fail?  Isn't it required to fail?

janaiyengar replied:
This is a reword of lines 3992-3996. I don't remember why this was written as a MAY earlier (as against a MUST, or just "fails"), but the new language is changing phrasing to "can" from "MAY". @martinthomson might remember why this isn't a failure requirement.

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,136 @@ The details of loss detection and congestion control are described in
+exceeds the total number of packets sent with each corresponding ECT codepoint.
+In particular, validation will fail when an endpoint receives a non-zero ECN
+count corresponding to an ECT codepoint that it never applied.  This check
+detects when packets are remarked to ECT(0) or ECT(1) in the network.
```
```suggestion
detects when packets are marked ECT(0) or ECT(1) by the network.
```

janaiyengar replied:
"Remarked" is important here, since it means that the endpoint marked packets as ECT(0) or ECT(1), and the network remarked it to one of those.

ianswett reviewed (changes requested):
Some suggestions, some of which I don't have suggestions for.

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,136 @@ The details of loss detection and congestion control are described in
-ECN counts.  A QUIC packet detected by a receiver as a duplicate does not
-affect the receiver's local ECN codepoint counts; see {{security-ecn}} for
-relevant security concerns.
+Use of ECN requires the receiving endpoint to read the ECN codepoint from an IP
```
Perhaps the following makes it easier? "requires the sending endpoint to write an ECN codepoint to IP headers and the receiving endpoint to read the ECN ..."

ianswett replied:
I was also unclear on what "Use of ECN" means, so I wanted to specify that as well.

martinthomson replied:
This section is about feedback only, so I think that it is OK to only talk about reading the markings.

gorryfair replied:
Can I suggest: /To enable the congestion controller to react to CE-marked packets, the sender needs to be able to set an ECT codepoint in the IP header and the receiver needs to be able to read the ECN field from an IP packet./
...i.e.,  removing /in order/ ... /ECN code point/ is the /ECN-field/ or name the code point value(s).

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,136 @@ The details of loss detection and congestion control are described in
-devices, each endpoint independently validates ECN counts and disables ECN if
-the path is not showing consistent support for ECN.
+An ECN count is not incremented when no QUIC packets from the received IP
+packet are processed. A QUIC packet detected by a receiver as a duplicate also
+does not increase an ECN count; see {{security-ecn}} for relevant security
```
Do we define what "processed" means? We probably should.

janaiyengar commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,136 @@ The details of loss detection and congestion control are described in
 
-* If all packets that were sent with the ECT(0) codepoint are eventually deemed
-  lost (Section 6 of {{QUIC-RECOVERY}}), validation is deemed to have failed.
+If an endpoint has cause to expect that IP packets with an ECT codepoint might
```
This isn't a departure from existing text -- see lines 3953-3956 in the old text.

gorryfair replied:
So, I’m going to chime-in here. Indeed it was in the “original” text, which is why I included this. However, we now have several people (me included) who doubt this is yet correct, so here are my thoughts:

(1)  I think it *is* important to be robust and describe how the protocol falls back in the corner case where the path drops IP packets with a non-zero ECN field - and we should very clearly say this.
(2) I think this can actually happen, but we should *not* make it sound like this often happens in general. I’m confident this is not “normal”.
(3) If we need to *measure* the path characteristic, using 10 packets seems a good test to me. However, as a fall back detection mechanism in a transport protocol this seems odd to stop using ECN after 10 packets when we expect that ECT is forwarded. Note: The text says {{ecn-ack}} says 10, the current text does not.

So.... This is a proposed change: We seem to be converging on clear text, would it be possible to remove the "10 packets" and revise this mechanism to say that the ECN counts need to increment, and pathological loss with no incrementing of counts is an indication that the sender should become “failed” for this connection? I think the text already virtually says this.

janaiyengar replied:
"Pathological loss" isn't going to be good enough as a mechanism, and there's a rathole waiting there to define precisely what "pathological" means. We chose 10 packets based on the IW, and also because we wanted to start somewhere and then change the mechanism as we learned more through experience.

I suggest that we don't muddle this particular (large) PR with an additional, potentially-contentious discussion about a mechanism change.

gorryfair replied:
I do like Jana's proposal to not muddle this particular (large) PR with this "issue". Would others agree to finalise the PR and then discuss this as a separate topic - afterwards?

mirjak replied:
I think we discussed this all ready once. I'm fine with 10 packet as it's the initial window (but maybe that's what we should say), however, given the result is the same I would rather state it as enable ECT marking and if all packet of your first flight are lost validation fails and you disable. Describing this specifically as probably makes it just sounds more complicated that it is.

ianswett replied:
The github diff didn't line up well, hence my confusion.  The way this is described still seems a bit unrealistic to me, but I can see this PR doesn't substantively change it, so I'm fine with that.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,136 @@ The details of loss detection and congestion control are described in
-enclosing IP packet or UDP datagram.
+On receiving a QUIC packet with an ECT(0), ECT(1) or CE codepoint, an
+ECN-enabled endpoint accesses the ECN codepoint from the enclosing IP packet and
+increases the corresponding ECT(0), ECT(1), or CE count, These ECN counts are
```
```suggestion
increases the corresponding ECT(0), ECT(1), or CE count. These ECN counts are
```

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,136 @@ The details of loss detection and congestion control are described in
+header are incremented once for each QUIC packet.
 
-### ECN Validation {#ecn-validation}
+For example, if one each of an Initial, 0-RTT, Handshake, and 1-RTT QUIC packet
```
That's true, except that this invalidates the other part of this paragraph...  This needs more thought.

ianswett replied:
Ok, happy to fix this in a separate PR if that makes sense?

ianswett replied:
I think this is technically legal, so I'm suggesting a note about how it's extremely unlikely.

gorryfair commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,136 @@ The details of loss detection and congestion control are described in
-associated IP header are incremented once for each QUIC packet, not per
-enclosing IP packet or UDP datagram.
+On receiving a QUIC packet with an ECT(0), ECT(1) or CE codepoint, an
+ECN-enabled endpoint accesses the ECN codepoint from the enclosing IP packet and
```
/ECN codepoint/ECN field/

martinthomson replied:
Is this throughout?  Because we have lots of use that should be "ECN field" if that is the case.

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,134 @@ The details of loss detection and congestion control are described in
+see {{ecn-validation}}.
 
 
 ### ECN Counts
```
```suggestion
### Reporting ECN Counts
```

martinthomson commented:
@larseggert @ianswett, you both hold blocking reviews on this.  There are enough comments here that I wasn't able to work out whether we'd addressed those comments.  Can you check and again and let us know if this is still blocking?

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,134 @@ The details of loss detection and congestion control are described in
+
+Use of ECN requires the receiving endpoint to read the ECN field from an IP
+packet, which is not possible on all platforms. If an endpoint does not
+implement ECN support or does not have access to received ECN field, it
```
```suggestion
implement ECN support or does not have access to received ECN fields, it
```

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,134 @@ The details of loss detection and congestion control are described in
+Even if an endpoint does not set an ECT field on packets it sends, the endpoint
+MUST provide feedback about ECN markings it receives, if these are accessible.
+Failing to report the ECN counts will cause the sender to disable use of ECN
+for packets to this receiver.
```
```suggestion
for this connection.
```

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,134 @@ The details of loss detection and congestion control are described in
-the path is not showing consistent support for ECN.
+Each packet number space maintains separate acknowledgement state and separate
+ECN counts.  Coalesced QUIC packets (see {{packet-coalesce}}) share the same IP
+header so the ECN counts are incremented once for each QUIC packet.
```
```suggestion
header so the ECN counts are incremented once for each coalesced QUIC packet.
```

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,134 @@ The details of loss detection and congestion control are described in
+
+For example, if one each of an Initial, Handshake, and 1-RTT QUIC packet are
+coalesced into a single UDP datagram, the ECN counts for all three packet
+number spaces will be incremented by one each.
```
```suggestion
number spaces will be incremented by one each, based on the ECN field of the single IP header.
```

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,134 @@ The details of loss detection and congestion control are described in
-first ten outgoing packets on a path, or for a period of three RTTs, whichever
-occurs first.
+If an endpoint has cause to expect that IP packets with an ECT codepoint might
+be dropped by faulty network element, the endpoint could set an ECT codepoint
```
```suggestion
be dropped by a faulty network element, the endpoint could set an ECT codepoint
```

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,134 @@ The details of loss detection and congestion control are described in
-successfully processed ACK frame. Any increase in ECN counts is validated
-based on the markings that were applied to packets that are newly
-acknowledged in the ACK frame.
+Erroneous application of ECN-CE markings by the network can result in degraded
```
```suggestion
Erroneous application of CE markings by the network can result in degraded
```

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,134 @@ The details of loss detection and congestion control are described in
-ECT(0) or ECT(1) codepoints set, ECN validation fails if ECN counts are not
-present in the ACK frame.  This check detects a network element that zeroes out
-ECN bits or a peer that is unable to access ECN markings.
+If an ACK frame newly acknowledges a packet that the endpoint sent setting
+either the ECT(0) or ECT(1) codepoint, ECN validation fails if the
```
```suggestion
If an ACK frame newly acknowledges a packet that the endpoint sent with
either the ECT(0) or ECT(1) codepoint set, ECN validation fails if the
```

larseggert commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,134 @@ The details of loss detection and congestion control are described in
+If an ACK frame newly acknowledges a packet that the endpoint sent setting
+either the ECT(0) or ECT(1) codepoint, ECN validation fails if the
+corresponding ECN counts are not present in the ACK frame. This check detects a
+network element that zeroes the ECN field or a peer that is unable to access
+ECN the markings.
```
```suggestion
network element that zeroes the ECN field or a peer that does not report 
ECN markings.
```

larseggert reviewed (approved):
I flagged some minor nits, but otherwise good to go.

ianswett commented on draft-ietf-quic-transport.md:
```diff
@@ -3874,145 +3874,134 @@ The details of loss detection and congestion control are described in
+
+For example, if one each of an Initial, Handshake, and 1-RTT QUIC packet are
+coalesced into a single UDP datagram, the ECN counts for all three packet
+number spaces will be incremented by one each.
```
```suggestion
number spaces will be incremented by one each.  Due to when keys are discarded, it's extremely unlikely a packet with all 4 types will arrive in the same datagram, but this forms a complete example.
```

martinthomson replied:
Isn't this based on old text?  This example lists three types only, which is possible (our stack can do three at the server in some cases).

ianswett reviewed (approved):
I made a suggestion on the 4 packets in a datagram issue that resolves it for me.  PTAL and otherwise I think this is fine.

martinthomson reviewed (commented):
Ian, this lost your addition (regarding the four types).  Do you think that we still need that?

ianswett commented:
> Ian, this lost your addition (regarding the four types). Do you think that we still need that?

Nope, your change is probably preferable.  I should have re-read the text more closely before posting the suggestion.
